Add optional ctl_connection_settings, force failure when can't get Table Metadata

kdparker opened the pull request:
Failure on replication handler when it can't get table metadata looks like this:
https://pb.yelpcorp.com/188948

This branch was tested using branch keegan_DATAPIPE-844-replhandler in the replication handler (where running it with this fix caused DATAPIPE-786 to be fixed without using the bandaid refresh runner)

julien-duponchelle committed:
Add more project using it

romuald committed:
Close connection on EOF packet

Since client might want to reconnect, reset connection status

julien-duponchelle committed:
Merge pull request #134 from romuald/eof-packet-reset-status

Close connection on EOF packet

jffifa committed:
bugfix for decoding values of columns in charset "utf8mb4"

julien-duponchelle committed:
Merge pull request #135 from jffifa/bugfix-for-utf8mb4

bugfix for decoding values of columns in charset "utf8mb4"

julien-duponchelle committed:
0.8 release

julien-duponchelle committed:
Add more contributors

julien-duponchelle committed:
PyMYSQL 0.7

mastak committed:
added intvar event

xiezhenye committed:
fix: missing update _next_seq_no

julien-duponchelle committed:
Merge pull request #139 from xiezhenye/fix_seq_no

fix: missing update _next_seq_no

Commit:
Report slave in SHOW SLAVE HOSTS

Signed-off-by: Arthur Gautier <baloo@gandi.net>

julien-duponchelle committed:
Merge pull request #143 from baloo/features/report-slave

Report slave in SHOW SLAVE HOSTS

Commit:
Support for slave_uuid

Allows mysql to disconnect stalled connection upon client reconnection.

Signed-off-by: Arthur Gautier <baloo@gandi.net>

julien-duponchelle committed:
Merge pull request #144 from baloo/features/slave_uuid

Support for slave_uuid

Commit:
binlogstream: improve slave report

Signed-off-by: Arthur Gautier <baloo@gandi.net>

julien-duponchelle committed:
Merge pull request #145 from baloo/fixup/slave_report

binlogstream: improve slave report

julien-duponchelle committed:
Merge pull request #147 from mastak/dev-intvar-event

added intvar event

Commit:
gtid: fixup parsing on mysql format

Mysql may return gtid joined by ',\n' separator:
 SELECT @@global.gtid_executed

Signed-off-by: Arthur Gautier <baloo@gandi.net>

Commit:
report-slave: Allow reporting of port if username not supplied

Signed-off-by: Arthur Gautier <baloo@gandi.net>

julien-duponchelle committed:
Merge pull request #148 from baloo/fixup/parsing-of-gtid

gtid: fixup parsing on mysql format

julien-duponchelle committed:
Merge pull request #149 from baloo/fixup/config-report-slave

report-slave: Allow reporting of port if username not supplied

Commit:
test run with mysql5.7

julien-duponchelle committed:
Merge pull request #150 from 26fe/mysql5.7

test run with mysql5.7

kdparker committed:
Add optional ctl_connection_settings to give seperate cluster to check changes in a given cluster rather than default, force failure on failure to get TableMetadata

justincinmd commented on pymysqlreplication/binlogstream.py:
```diff
@@ -56,14 +57,16 @@ def __init__(self, connection_settings, server_id, resume_stream=False,
         self.__connected_ctl = False
         self.__resume_stream = resume_stream
         self.__blocking = blocking
+        self._ctl_connection_settings = ctl_connection_settings
```
Do you need to set a charset for this too?

justincinmd commented on pymysqlreplication/row_event.py:
```diff
@@ -55,7 +55,7 @@ def __init__(self, from_packet, event_size, table_map, ctl_connection, **kwargs)
 
         if len(self.columns) == 0:  # could not read the table metadata, probably already dropped
-            self.complete = False
+            raise TableMetadataUnavailableError("Table Metadata could not be read for {}".format(self.table))
```
Maybe this should be optional?  I absolutely want it, but the maintainer didn't - I think it's much more correct to raise an exception, but if we want to contribute back upstream, we probably shouldn't force the issue.

kdparker committed:
Made failing table metadata unavailable error an optional parameter of BinLogStreamReader, added charset for ctl_connection_settings

julien-duponchelle committed:
Add more related projects

Commit:
setup travis to run test on multiple mysql version (5.6 and 5.7)

julien-duponchelle committed:
Merge pull request #153 from 26fe/test-multiple-mysql-version

setup travis to run test on multiple mysql version (5.6 and 5.7)

Commit:
Allow to specify a pymysql wrapper or subclass

This will allow to implement custom transports (tls transport and not mysql own
starttls equivalent)

example usage:
```python
import socket

import pymysql.connections
from pymysql.constants import CLIENT

class TLSConnection(pymysql.connections.Connection):
    def __init__(self, *args, **kwargs):
        kwargs['defer_connect'] = True

        super(TLSConnection, self).__init__(*args, **kwargs)
        self.ssl = False
        self.client_flag &= ~CLIENT.SSL
        sock = socket.create_connection((self.host, self.port), self.connect_timeout)
        self.host_info = "socket %s:%d" % (self.host, self.port)
        sock.setsockopt(socket.IPPROTO_TCP, socket.TCP_NODELAY, 1)
        sock = self.ctx.wrap_socket(sock, server_hostname=self.host)

        self.connect(sock)

mysql_settings={
    'host': '1.2.3.4',
    'user': 'acme',
    'password': 'hackme',
    'port': 3306,
    'charset': 'utf8mb4',
    'db': '',
    'ssl': {
      'ca': 'root.crt',
      'cert': 'crt',
      'key':  'key',
      'cipher': 'ECDHE-RSA-AES256-GCM-SHA384'
    }
}

from pymysqlreplication import BinLogStreamReader
stream = BinLogStreamReader(connection_settings=mysql_settings, pymysql_wrapper=TLSConnection)

```

Signed-off-by: Arthur Gautier <baloo@gandi.net>

julien-duponchelle committed:
Merge pull request #156 from baloo/features/custom_tranport

Allow to specify a pymysql wrapper or subclass

julien-duponchelle committed:
0.9

julien-duponchelle committed:
Add me as contributor

asottile committed:
unittest2 is only a test requirement

Commit:
documentation: fixup warning about pymysql0.7

Fixed by 12fa276065a3f82f7659e70fb8790a0b561674e6

Tested-by: Arthur Gautier <baloo@gandi.net>
Signed-off-by: Arthur Gautier <baloo@gandi.net>

baloo committed:
Merge pull request #158 from asottile/unittest2_test_requirement

unittest2 is only a test requirement

dveeden committed:
Fix the example: Add required server_id

Otherwise this happens:
```
Traceback (most recent call last):
  File "./test.py", line 6, in <module>
    stream = BinLogStreamReader(connection_settings = mysql_settings)
TypeError: __init__() missing 1 required positional argument: 'server_id'
```

bjoernhaeuser committed:
Merge pull request #160 from dveeden/fix_readme_example

Fix the example: Add required server_id

mastak committed:
fixed count of fetched events

julien-duponchelle committed:
Merge pull request #161 from mastak/tests-typo

fixed count of fetched events

julien-duponchelle committed:
Add mymongo

toadzhou committed:
Update mysql_to_logstash.py

Repair the parameter error

julien-duponchelle committed:
Merge pull request #166 from toadzhou/patch-1

Update mysql_to_logstash.py

abrarsheikh committed:
encode mysql time as python datetime.timedelta

bjoernhaeuser committed:
Fix build by updating to mysql 5.7.14

baloo committed:
Merge pull request #170 from noplay/fix-build

Fix build by updating to mysql 5.7.14

Commit:
Merge pull request #169 from abrarsheikh/master

encode mysql time as python datetime.timedelta

Signed-off-by: Arthur Gautier <baloo@gandi.net>

julien-duponchelle committed:
Add pg_chameleon to the list of project using the lib

Commit:
Decode gtid from network packet

Signed-off-by: Arthur Gautier <baloo@gandi.net>

Commit:
doc: Running mysql in docker

Signed-off-by: Arthur Gautier <baloo@gandi.net>

julien-duponchelle committed:
Merge pull request #175 from baloo/baloo/doc/tests-in-docker

doc: Running mysql in docker

julien-duponchelle committed:
Merge pull request #174 from baloo/baloo/features/decode-gtid-from-packet

Decode gtid from network packet

premsantosh commented on pymysqlreplication/binlogstream.py:
```diff
@@ -35,14 +35,16 @@ class BinLogStreamReader(object):
                  only_tables=None, only_schemas=None,
-                 freeze_schema=False, skip_to_timestamp=None):
+                 freeze_schema=False, skip_to_timestamp=None,
+                 fail_on_table_metadata_unavailable=False):
```
Is this a temporary flag? And why is it defaulted to false? When do we not want to get alerted?

kdparker replied:
I think the intention was we eventually wanted to merge this in to the overall master branch of the library, and to do so we want to keep existing functionality exactly the same (the corresponding branch in replhandler sets this to True).

Do we want to make this default to True?

premsantosh replied:
Aah I see, that makes sense. If we are setting it to true from our side then it should be fine.

premsantosh commented on pymysqlreplication/exceptions.py:
```diff
@@ -0,0 +1,2 @@
+class TableMetadataUnavailableError(Exception):
+    pass
```
Would be nice to have a little context with this. Maybe something like:

```
class TableMetadataUnavailableError(Exception):
  def __init__(self, table_name, cluster):
    Exception.__init__(self, "Unable to find metadata for table {t} in cluster {c}".format(..))
```

kdparker commented:
Something else we may want to consider is migrating this to another user/merge it in to the noplay master (can anyone here actually merge these changes?)

kdparker committed:
Fill out TableMetadataUnavailableError to receive a table instead of it being a generic exception

kdparker committed:
Merge branch 'master' of https://github.com/noplay/python-mysql-replication

kdparker committed:
Add test for metadata unavailable failure, fix metadata unavailable exception builder

kdparker committed:
Format fix for python 2.6 support

kdparker committed:
Merge pull request #1 from kdparker/optional_ctl_connection_settings

Merge noplay's changes in to optional_ctl_connection_settings, add test for metadata unavailable error

kdparker committed:
Added tests for ctl_connection_settings, removed vim files from gitignore, made default charset applying simpler, fixed travis for new mysql servers necessary for ctl_connection_settings test (including removing usage of mysql sandbox in mysql57 tests to bring it in line with what's going on in mysql56 tests)

kdparker commented:
Not using this branch anymore, closing
